### PR TITLE
mon,mds: use per-MDS compat to inform replacement

### PR DESCRIPTION
**Pull request by batrick**
Fixes: https://tracker.ceph.com/issues/49720

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

**Comment by batrick** [comment]
(added a doc commit and rebased)

**Comment by tchaikov** [comment]
```
flake8 run-test: commands[0] | flake8 --select=F,E9 --exclude=venv,.tox
./tasks/mds_pre_upgrade.py:6:1: F401 'time' imported but unused
./tasks/mds_pre_upgrade.py:25:5: F841 local variable 'status' is assigned to but never used
./tasks/cephfs/test_admin.py:340:13: F821 undefined name 'fs2'
ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/qa/.tox/flake8/bin/flake8 --select=F,E9 --exclude=venv,.tox (exited with code 1)
```

**Comment by batrick** [comment]
Thanks @tchaikov , fixed!

File: src/mds/FSMap.cc
Comment (gregsfortytwo) [inline src/mds/FSMap.cc:762]: This doesn't seem correct; an MDS might support features which the existing FS can't handle, but that doesn't mean the MDS can't write to the existing FS while using an older mode, does it?

File: src/mds/MDSMap.cc
Comment (gregsfortytwo) [inline src/mds/MDSMap.cc:614]: It doesn't look to me like we're initializing ```compat``` with a good default value in the constructor or here when we decode an older MDSMap?

File: src/mon/MonCommands.h
Comment (gregsfortytwo) [inline src/mon/MonCommands.h:419]: I guess some version of compat modification existed before so maybe I'm just behind the times, but what's the use case for these commands?

File: doc/cephfs/upgrading.rst
Comment (gregsfortytwo) [inline doc/cephfs/upgrading.rst:41]: I think this needs to be louder and clearer, so that people don't leave old versions of the code running without realizing that it won't help them. As is it just reads like "yeah you should upgrade the other daemons for a better experience".

File: src/mon/FSCommands.cc
Comment (batrick) [reply src/mon/FSCommands.cc:395]: It's maintained by the `inline_data_enabled` setting on the `MDSMap`. We can't really know if a file system "ever" had it turned on. That's not really saved anywhere.

As it stands with this change, all file systems will add that incompat feature because all MDS (Pacific onwards) will advertise it.

File: src/mds/Beacon.cc
Comment (batrick) [reply src/mds/Beacon.cc:237]: conservative

File: src/mds/FSMap.h
Comment (batrick) [reply src/mds/FSMap.h:193]: This is a temporary restriction, as I see it. It's a stop-gap between turning off all MDS (except rank 0 of one file system, our current recommendation) and enabling a rolling upgrade without reducing max_mds.

> Why not just keep the intersection of MDS features as the filesystem compat set?

You mean instead of storing the CompatSet in the MDSMap? Because those features are sometimes used to indicate on-disk format changes (snapv2 for example).

File: src/mds/FSMap.cc
Comment (batrick) [reply src/mds/FSMap.cc:762]: In theory but the MDS is not coded to handle that. snapv2 comes to mind, again.

File: src/mds/MDSMap.cc
Comment (batrick) [reply src/mds/MDSMap.cc:614]: Hmm, you're right. Thanks. I'll fix it.

File: src/mon/MonCommands.h
Comment (batrick) [reply src/mon/MonCommands.h:419]: Right now, just testing. I've been considering if I should flag them `HIDDEN`.

File: doc/cephfs/upgrading.rst
Comment (batrick) [reply doc/cephfs/upgrading.rst:41]: I'm not sure we're on the same page. This note is saying that they should upgrade standbys **first**. Of course, they should upgrade all the daemons. So I'm not sure what you mean "leave old versions of the code running".

File: src/mds/FSMap.cc
Comment (tchaikov) [inline src/mds/FSMap.cc:133]:
```
2021-04-02T19:55:48.100 INFO:tasks.cephfs_test_runner:test_full_health (tasks.mgr.dashboard.test_health.HealthTest) ... FAIL
2021-04-02T19:55:48.101 INFO:tasks.cephfs_test_runner:
2021-04-02T19:55:48.101 INFO:tasks.cephfs_test_runner:======================================================================
2021-04-02T19:55:48.102 INFO:tasks.cephfs_test_runner:FAIL: test_full_health (tasks.mgr.dashboard.test_health.HealthTest)
2021-04-02T19:55:48.102 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2021-04-02T19:55:48.102 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2021-04-02T19:55:48.103 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_3d0f85a6c3046372f708a5f2893231db573d2f75/qa/tasks/mgr/dashboard/helper.py", line 480, in assertSchema
2021-04-02T19:55:48.103 INFO:tasks.cephfs_test_runner:    return _validate_json(data, schema)
2021-04-02T19:55:48.103 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_3d0f85a6c3046372f708a5f2893231db573d2f75/qa/tasks/mgr/dashboard/helper.py", line 710, in _validate_json
2021-04-02T19:55:48.104 INFO:tasks.cephfs_test_runner:    for key, sub_schema in schema.sub_elems.items()
2021-04-02T19:55:48.104 INFO:tasks.cephfs_test_runner:tasks.mgr.dashboard.helper._ValError: In `input['fs_map']`: missing keys: {'compat'}
2021-04-02T19:55:48.104 INFO:tasks.cephfs_test_runner:
2021-04-02T19:55:48.104 INFO:tasks.cephfs_test_runner:During handling of the above exception, another exception occurred:
2021-04-02T19:55:48.105 INFO:tasks.cephfs_test_runner:
2021-04-02T19:55:48.105 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2021-04-02T19:55:48.105 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_3d0f85a6c3046372f708a5f2893231db573d2f75/qa/tasks/mgr/dashboard/test_health.py", line 273, in test_full_health
2021-04-02T19:55:48.106 INFO:tasks.cephfs_test_runner:    self.assertSchema(data, schema)
2021-04-02T19:55:48.106 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_3d0f85a6c3046372f708a5f2893231db573d2f75/qa/tasks/mgr/dashboard/helper.py", line 482, in assertSchema
2021-04-02T19:55:48.106 INFO:tasks.cephfs_test_runner:    self.assertEqual(data, str(e))
2021-04-02T19:55:48.107 INFO:tasks.cephfs_test_runner:AssertionError: {'health': {'status': 'HEALTH_OK', 'check[322522 chars]: 0}} != "In `input['fs_map']`: missing keys: {'compat'}"
2021-04-02T19:55:48.107 INFO:tasks.cephfs_test_runner:
```
see https://pulpito.ceph.com/kchai-2021-04-02_15:50:13-rados-wip-kefu-testing-2021-04-02-1506-distro-basic-gibba/6017625/

File: src/mds/MDSMap.cc
Comment (batrick) [reply src/mds/MDSMap.cc:614]: Fixed

File: src/mds/FSMap.cc
Comment (batrick) [reply src/mds/FSMap.cc]: Thanks @tchaikov . I reverted the name change for the JSON.

File: src/mon/FSCommands.cc
Comment (batrick) [reply src/mon/FSCommands.cc:31]: Fixed

File: src/mds/Beacon.cc
Comment (batrick) [reply src/mds/Beacon.cc:237]: I pulled this out into a separate commit with explanation.

**Comment by batrick** [comment]
ping @gregsfortytwo @ceph/cephfs

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
@ajarr @joscollin plase review.

**Comment by batrick** [comment]
jenkins test api

**Comment by tchaikov** [comment]
```
c++: fatal error: Killed signal terminated program cc1plus
compilation terminated.
```

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test make check arm64

File: src/mon/FSCommands.cc
Comment (batrick) [reply src/mon/FSCommands.cc:714]: We use a mix of present progressive and past tense; I don't think this matters.

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-July-30

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-July-30

**Comment by batrick** [comment]
Ready to merge after approval.

File: src/mds/FSMap.cc
Comment (ajarr) [reply src/mds/FSMap.cc:774]: Should this be, !upgradeable || !fs.mds_map.compat.writeable(info.compat) ?

**Review by ajarr — COMMENTED** [review]
Most looks good to me. I've a question, and a suggested doc change.

File: src/mds/FSMap.cc
Comment (batrick) [reply src/mds/FSMap.cc:774]: No. If the file system is not upgradeable but `fs.mds_map.compat.writeable(info.compat) == true`, we still want to be able to promote that standby.

File: doc/cephfs/upgrading.rst
Comment (batrick) [reply doc/cephfs/upgrading.rst:53]: Thanks, fixed!

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test api

**Review by joscollin — APPROVED** [review]

**Review by ajarr — APPROVED** [review]

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test api

**Pull request merged**
